temporal/fx.go

```diff
@@ -514,54 +520,53 @@ func ApplyClusterMetadataConfigProvider(
 		resp, err := clusterMetadataManager.GetClusterMetadata(&persistence.GetClusterMetadataRequest{
 			ClusterName: clusterName,
 		})
-		if err != nil {
+		switch err.(type) {
+		case nil:
+			// Verify current cluster metadata
+			if clusterName == clusterData.CurrentClusterName {
+				var persistedShardCount = resp.HistoryShardCount
+				if config.Persistence.NumHistoryShards != persistedShardCount {
+					logger.Error(
+						mismatchLogMessage,
+						tag.Key("persistence.numHistoryShards"),
+						tag.IgnoredValue(config.Persistence.NumHistoryShards),
+						tag.Value(persistedShardCount))
+					config.Persistence.NumHistoryShards = persistedShardCount
+				}
+			}
+		case *serviceerror.NotFound:
+			if clusterName == clusterData.CurrentClusterName {
+				// Case 3: data exists in cluster metadata but not in cluster metadata info. Back fill data.
+				oldClusterMetadata, err := clusterMetadataManager.GetClusterMetadataV1()
+				if err != nil {
+					return config.ClusterMetadata, config.Persistence, fmt.Errorf("error while getting old cluster metadata: %w", err)
+				}
+				applied, err = clusterMetadataManager.SaveClusterMetadata(&persistence.SaveClusterMetadataRequest{
+					ClusterMetadata: persistencespb.ClusterMetadata{
+						HistoryShardCount:        oldClusterMetadata.HistoryShardCount,
+						ClusterName:              oldClusterMetadata.ClusterName,
+						ClusterId:                oldClusterMetadata.ClusterId,
+						VersionInfo:              oldClusterMetadata.VersionInfo,
+						IndexSearchAttributes:    oldClusterMetadata.IndexSearchAttributes,
+						ClusterAddress:           clusterInfo.RPCAddress,
+						FailoverVersionIncrement: clusterData.FailoverVersionIncrement,
+						InitialFailoverVersion:   clusterInfo.InitialFailoverVersion,
+						IsGlobalNamespaceEnabled: clusterData.EnableGlobalNamespace,
+						IsConnectionEnabled:      clusterInfo.Enabled,
+					}})
+				if err != nil || !applied {
+					return config.ClusterMetadata, config.Persistence, fmt.Errorf("error while backfiling cluster metadata: %w", err)
+				}
+			} else {
+				return config.ClusterMetadata,
+					config.Persistence,
+					fmt.Errorf("error while fetching metadata from cluster %s: %w", clusterName, err)
+			}
+		default:
 			return config.ClusterMetadata,
 				config.Persistence,
 				fmt.Errorf("error while fetching metadata from cluster %s: %w", clusterName, err)
 		}
-
-		if clusterName == resp.GetClusterName() {
```

Comment:
why are these validation no longer needed?

Reply:
It checks the history shard count. I removed the rest because there will be a metadata cache to load everything from DB. I am fine to keep the checks and print out the error messages.